src/creatwth.cpp

```diff
@@ -180,16 +180,6 @@ static PBYTE FindAndAllocateNearBase(HANDLE hProcess, PBYTE pbBase, DWORD cbAllo
 
         PBYTE pbAddress = (PBYTE)(((DWORD_PTR)mbi.BaseAddress + 0xffff) & ~(DWORD_PTR)0xffff);
```

Comment:
I'm not convinced that the problem is that VirtualQueryEx is rounding pbBase down to the nearest page boundary.  I think there may be another issue going on here.

If the problem is VirtualQueryEx rounding down to the nearest page boundary, then line 181 should fix this. It rounds pbAddress up to the nearest 64K boundary.  Assuming that pages are less than 64K in size, pbAddress would be at or above pbBase.   (I don't recall whether Windows x64 supports large virtual page sizes (2Mbyte) that are supported by the x64 architecture, in which case this reasoning is flawed).

I suspect the problem is that mbi.BaseAddress may be rounded down even further.  The only guarantee from VirtualQueryEx is that lpAddress (the address for which we're querying information about) is within the memory range starting at mbi.BaseAddress.   Suppose  that mbi.BaseAddress is aligned to a memory manager allocation granularity and <  pbBase.  Then line 181 doesn't do anything, and the existing check fails.

Of course, mbi.RegionSize could be even lower, depending on what the memory manager decided to do.   In this case, pbAlloc could be < pbBase.

Reply:
Independing of all the rounding concerns, the original check, that I put in, is overly strict in the sense that Austin is experiencing.

The allocation and certain bytes within it must be within 4GB of the start of the image.
The check however checks for roughly 4GB from the end of the image -- it doesn't compute the end of the image very well imho, but it tries. It uses end of the image, because that is where it tries to start allocating from.

This was my mistake, a few years ago.

Historically the check wasn't even there.

And on desktop it rarely mattered -- for the CreateProcessWithDll case -- it'd always succeed.
I added the check, because on some Windows variants, sections (images) and virtualmemory occupy separate and far apart ranges (more than 4GB apart), and the address "successfully" allocated here was never within 4GB. And it'd get silently truncated, and then crash in ntdll.dll using the truncated offset.

Reply:
Hi David, 

My understanding of VirtualQueryEx's behavior is based on [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa366907(v=vs.85).aspx), which says that "[lpAddress] is rounded down to the next page boundary". This description matches the behavior I saw while debugging the issue, but I wouldn't know if it can act differently in other cases.

> If the problem is VirtualQueryEx rounding down to the nearest page boundary, then line 181 should fix this. It rounds pbAddress up to the nearest 64K boundary. Assuming that pages are less than 64K in size, pbAddress would be at or above pbBase.

I believe you're correct, unless pbBase lies in the first page after a 64K boundary. In that case, VirtualQueryEx would round it down to the 64K boundary, and line 181 would assign pbAddress to the same 64K boundary. pbAddress would then be less than pbBase, making us fail the old 4GB checks. 

Indeed, this is exactly what I see when repro'ing this issue. I've copied the relevant code below, annoted with extra bolded comments describing misc variable values when the issue occurs. 

It sounds like I may have missed a third precondition for pbBase in my PR comment: that is, pbBase needs to lie in the first page after a 64K boundary.

Thanks for your help.

<pre>
static PBYTE FindAndAllocateNearBase(HANDLE hProcess, PBYTE pbBase, DWORD cbAlloc)
{
    <b>//
    // pbBase = 0x00007ff79a920800
    //</b>
 
    MEMORY_BASIC_INFORMATION mbi;
    ZeroMemory(&mbi, sizeof(mbi));
 
    PBYTE pbLast = pbBase;
    for (;; pbLast = (PBYTE)mbi.BaseAddress + mbi.RegionSize) {
 
        ZeroMemory(&mbi, sizeof(mbi));
        if (VirtualQueryEx(hProcess, (PVOID)pbLast, &mbi, sizeof(mbi)) == 0) {
            if (GetLastError() == ERROR_INVALID_PARAMETER) {
                break;
            }
            DETOUR_TRACE(("VirtualQueryEx(%p) failed: %d\n",
                          pbLast, GetLastError()));
            break;
        }
 
        <b>
        //
        // According to MSDN, VirtualQueryEx rounds pbLast down to its nearest page boundary.
        // As a result, mbi.BaseAddress = 0x00007ff79a920000.
        // </b>

        // Usermode address space has such an unaligned region size always at the
        // end and only at the end.
        //
        if ((mbi.RegionSize & 0xfff) == 0xfff) {
            break;
        }
 
        // Skip anything other than a pure free region.
        //
        if (mbi.State != MEM_FREE) {
            continue;
        }
       
        <b>
        //
        // Since 0x00007ff79a920000 is 64KB-aligned, the following line just sets pbAddress = 0x00007ff79a920000
        //</b>
        PBYTE pbAddress = (PBYTE)(((DWORD_PTR)mbi.BaseAddress + 0xffff) & ~(DWORD_PTR)0xffff);
 
#ifdef _WIN64
        // The distance from pbBase to pbAddress must fit in 32 bits.
        //
        const size_t GB4 = ((((size_t)1) << 32) - 1);
        if ((size_t)(pbAddress - pbBase) > GB4) {
            <b>
            //
            // The above statement is true. pbAddress = 0x00007ff79a920000 and pbBase = 0x00007ff79a920800,
            // so “pbAddress – pbBase” is a very large number when casted to size_t.
            // FindAndAllocateNearBase returns NULL as a result, and DetourUpdateProcessWithDll fails.
            //</b>
            DETOUR_TRACE(("FindAndAllocateNearBase(1) failing due to distance >4GB %p\n", pbAddress));
            return NULL;
        }
#endif
                …
}
</pre>